sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2398,8 +2398,14 @@ class Dataset[T] private[sql](
   /**
    * Returns a new Dataset by adding columns or replacing the existing columns that has
    * the same names.
+   *
+   * `column`'s expression in `cols` must only refer to attributes supplied by this Dataset.
+   * It is an error to add columns that refers to some other Dataset.
+   *
+   * @group untypedrel
+   * @since 3.2.0
    */
-  private[spark] def withColumns(colNames: Seq[String], cols: Seq[Column]): DataFrame = {
+  def withColumns(colNames: Seq[String], cols: Seq[Column]): DataFrame = {
```

Comment:
If we're adding them, I think we should change the signature to either map or a list of tuple

Reply:
I'd like to add Map like `colsMap`, it's more readable, I will do it in next PR.